observatory-platform/observatory/platform/bigquery.py

```diff
@@ -294,6 +298,71 @@ def bq_find_schema(
     return None
 
 
+def bq_generate_schema_from_data(data_path: str) -> List[dict]:
+    """
+    Generates a Biguqery style schema based on a *.jsonl.gz data file.
+
+    :param expected_schema_path: Path to the expected schema of the data.
+    :param data_path: Path to the *.jsonl.gz data file.
+    :param check_data_types: Optional, set to True to check the data type for expected vs generated.
+    :return: True if the expected and generated schemas match.
+    """
+
+    with CliRunner().isolated_filesystem() as temp_file_path:
+        generator = SchemaGenerator()
+        random_id = observatory_environment.random_id()
+        generated_schema_path = os.path.join(temp_file_path, f"generated_schema_{random_id}.json")
+        with gzip.open(data_path, "rb") as f_in, open(generated_schema_path, "w") as f_out:
+            generator.run(input_file=f_in, output_file=f_out, schema_map=None)
+        with open(generated_schema_path, "r") as f_in:
+            generated = json.load(f_in)
+
+    return generated
+
+
+def bq_compare_schemas(expected: List[dict], actual: List[dict], check_types_match: Optional[bool] = False) -> bool:
+    """Compare two Bigquery schemas for if they have the same fields and/or data types.
+
+    :param expected: the expected schema.
+    :param actual: the actual schema.
+    :check_types_match: Optional, if checking data types of fields is required.
+    :return: whether the expected and actual match.
+    """
+
+    expected.sort(key=lambda c: c["name"], reverse=False)
+    actual.sort(key=lambda c: c["name"], reverse=False)
+
+    exp_names = [field_def["name"] for field_def in expected]
+    act_names = [field_def["name"] for field_def in actual]
+
+    if len(exp_names) != len(act_names):
+        logging.error("Fields do not match:")
+        logging.error(f"Only in expected: {set(exp_names) - set(act_names)}")
+        logging.error(f"Only in actual: {set(act_names) - set(exp_names)}")
+        return False
+
+    # Check data types of fields
+    if check_types_match:
+        for exp_field, act_field in zip(expected, actual):
+            if exp_field["type"] != act_field["type"]:
+                logging.error(
+                    f"Field types do not match for field  '{exp_field['name']}' ! Actual: {act_field['type']} vs Expected: {exp_field['type']}"
+                )
+            all_matched = False
+
+    # Check for sub-fields within the schema.
+    all_matched = True
+    for exp_field, act_field in zip(expected, actual):
+        # Ignore the "mode" and "description" definitions in fields as they are not required for check.
+        diff = DeepDiff(exp_field, act_field, ignore_order=True, exclude_regex_paths=r"\s*(description|mode)")
```

Comment:
Maybe ignore_order=True and exclude_regex_paths should be keyword parameters in case they need changing.